rlclientlib/utility/context_helper.cc

```diff
@@ -61,120 +62,83 @@ namespace reinforcement_learning { namespace utility {
-    std::string parse_error;
-    count = get_array_count(context, multi, trace, status, parse_error);
-    if (!parse_error.empty()) {
-      RETURN_ERROR_LS(trace, status, json_parse_error) << parse_error;
+      ++_level;
+      return true;
     }
 
-    if (count > 0) {
-      return error_code::success;
+    bool EndObject(rj::SizeType memberCount)
+    {
+      --_level;
+
+      if((_is_multi | _is_slots) && _level == 1 && _array_level == 1) {
+        size_t item_end = _is.Tell() - 1;
+        if(_is_multi)
+          _info.actions.push_back(std::make_pair(_item_start, item_end));
+        if(_is_slots)
+          _info.slots.push_back(std::make_pair(_item_start, item_end));
+      }
+      return true;
     }
 
-    RETURN_ERROR_LS(trace, status, json_no_actions_found);
-  }
-
-  /**
-   * \brief Get the number of slots found in context json string.  Slots should be in an array
-   * called _slots under the root name space.  {_slots:[{"name1":"val1"},{"name1":"val1"}]}
-   *
-   * \param count   : Return value passed in as a reference.
-   * \param context : String with context json
-   * \param trace   : Pointer to the trace logger
-   * \param status  : Pointer to api_status object that contains an error code and error description in
-   *                  case of failure
-   * \return  error_code::success if there are no errors.  If there are errors then the error code is
-   *          returned.
-   */
-  int get_slot_count(size_t& count, const char *context, i_trace* trace, api_status* status) {
-    std::string parse_error;
-    count = get_array_count(context, slots, trace, status, parse_error);
-    if (!parse_error.empty()) {
-      RETURN_ERROR_LS(trace, status, json_parse_error) << parse_error;
+    bool StartArray()
+    {
+      ++_array_level;
+      return true;
     }
 
-    if ( count > 0 ) {
-      return error_code::success;
+    bool EndArray(rj::SizeType elementCount)
+    {
+      --_array_level;
+      return true;
     }
+  };
 
-    RETURN_ERROR_LS(trace, status, json_no_slots_found);
-  }
-
-  /**
-   * \brief Validate that the _multi and _slots entries exist in the context json string,
-   * and that _multi comes before _slots
-   * 
-   * \param context : String with context json
-   * \param trace   : Pointer to the trace logger
-   * \param status  : Pointer to api_status object that contains an error code and error description in
-   *                  case of failure
-   * \return  error_code::success if there are no errors.  If there are errors then the error code is
-   *          returned.
-   */
-  int validate_multi_before_slots(const char *context, i_trace* trace, api_status* status)
+  int get_context_info(const char *context, ContextInfo &info, i_trace* trace, api_status* status)
   {
-    auto slots_pos = strstr(context, "_slots");
-    auto multi_pos = strstr(context, "_multi");
+    std::string copy(context);

```

Comment:
is it required to make copy here?

Reply:
Yes, I'm using kParseInsituFlag which modified the string.
I did some testing and it's measurably faster even with the copy.